internal/core/api.rs

```diff
@@ -11,6 +11,20 @@ use alloc::rc::Rc;
 use crate::component::ComponentVTable;
 use crate::window::WindowRc;
 
+pub use crate::lengths::LogicalPx;
+pub use crate::lengths::PhysicalPx;
+
+/// Convenience alias for representing a size in logical pixels (before applying a window scale factor).
+pub type LogicalSize = euclid::Size2D<f32, LogicalPx>;
+/// Convenience alias for representing a size in physical pixels.
+pub type PhysicalSize = euclid::Size2D<i32, PhysicalPx>;
+/// Convenience alias for representing a point in logical pixels (before applying a window scale factor).
+pub type LogicalPoint = euclid::Point2D<f32, LogicalPx>;
+/// Convenience alias for representing a point in physical pixels.
+pub type PhysicalPoint = euclid::Point2D<i32, PhysicalPx>;
+/// Convenience alias for a scale that can be used with euclid's types to convert between logical and physical pixels.
+pub type LogicalToPhysicalScale = euclid::Scale<f32, LogicalPx, PhysicalPx>;
```

Comment:
I wonder if `f32` and `i32` are the correct primitive data types here. i32 makes sense as physical pixel unit, but it means that after applying a the scale the result is an Euclid Point/Size that's `f32`, so conversion to an actual PhysicalPoint/PhysicalSize alias requires a `.cast()`.

Another option would be to say that the logical coordinate grid is also pixels in the API. But I guess that means scale factors would have to be integral values.

And finally, using `f32` everywhere is also a bit weird - in the light of actual pixels on the screen being discrete.

Reply:
I think it sucks that rust can't do basic arithmetic between different type like integer and floats.

Now, what i'm concerned about is that the software renderer uses u16 for physical lenght and now we have two PhysicalLenghth.

I wonder if it is not simpler to just keep the type as `euclid::Point2D<i32, PhysicalPx>` then it's clear what it is.

But yes, that means we need to re-export euclid in a module.

Reply:
We could do that, but we also make our API less expressive then. Now:

`app.window().set_position(slint::PhysicalPoint::new(.., ..));`

and without the alias:

`app.window().set_position(slint::euclid::Point2D::new(.., ..));`

The tag is inferred.

That said, we can't make everything totally water-tight. The expected type needs to be documented with `set_position` anyway and the tag is definitely there when people start to store these values in variables. And then the type "safety" kicks in.

Reply:
(Done)